### PR TITLE
add otel.kind and otel.status_code to follow otel spec

### LINKED ISSUE #909: Span kinds should be set correctly
Currently spans do not have their kind set.
We should follow the Otel spec.
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind

### PR DESCRIPTION
**Pull request by bnjjj**
close #909 

I added `SpanKind` and `SpanStatusCode` to be compliant with the otel spec and also follow the same rules than the Gateway. 
I also added a `.instrument(...)` to be able to see the parsing of the query in our main trace, before it was considered as a standalone span because it's spawned with a `spawn_blocking`

### CODE REVIEW

**Review by BrynCooke — COMMENTED** [review]
Is it possible to have all the spans in a unit test? I'd love to see some client spans in the test snapshots.

File: apollo-router-core/src/query_cache.rs [hunk 1]
```diff
@@ -18,7 +21,8 @@ struct QueryCacheResolver {
 impl CacheResolver<String, Option<Arc<Query>>> for QueryCacheResolver {
     async fn retrieve(&self, key: String) -> Result<Option<Arc<Query>>, CacheResolverError> {
         let schema = self.schema.clone();
-        let query_parsing_future = tokio::task::spawn_blocking(move || Query::parse(key, &schema));
+        let query_parsing_future = tokio::task::spawn_blocking(move || Query::parse(key, &schema))
+            .instrument(info_span!("parse_query", "otel.kind" = %SpanKind::Internal));
```
Comment (garypen) [inline apollo-router-core/src/query_cache.rs:25]: I don't know if this is a problem or not, but won't we now have two spans called "parse_query"?

File: apollo-router-core/src/spec/query.rs [hunk 2]
```diff
@@ -84,7 +85,7 @@ impl Query {
         }
     }
 
-    #[tracing::instrument(skip_all, level = "info" name = "parse_query")]
+    #[tracing::instrument(skip_all, level = "info" name = "parse_query", fields(otel.kind=%SpanKind::Internal) )]
```
Comment (garypen) [inline apollo-router-core/src/spec/query.rs:88]: This being the other "parse_query" span...

File: apollo-router-core/src/query_cache.rs (hunk 1 above)
Comment (garypen) [reply apollo-router-core/src/query_cache.rs:25]: I don't know either. Maybe discuss tomorrow?

**Comment by bnjjj** [comment]
@BrynCooke the subgraph trace with a client kind should already be in our snapshotted spans but I spotted a bug in test-span (too aggressive on filtering spans) then I will write a fix next week on test-span and it should be ok. I'm on it :)

**Comment by bnjjj** [comment]
as discussed in our meeting, I will open a new PR with the test-span fix

File: apollo-router/src/axum_http_server_factory.rs [hunk 3]
```diff
@@ -511,10 +523,88 @@ fn prefers_html(accept_header: &HeaderValue) -> bool {
 }
 
 #[derive(Clone)]
-struct PropagatingMakeSpan(DefaultMakeSpan);
+struct PropagatingMakeSpan {
+    level: Level,
+}
+
+impl PropagatingMakeSpan {
```
Comment (BrynCooke) [inline apollo-router/src/axum_http_server_factory.rs:530]: Suggest not parameterizing this. We will only ever want Info.
This removes the need for the macro stuff later.

File: apollo-router/tests/snapshots/integration_tests__traced_basic_request.snap [hunk 4]
```diff
@@ -18,46 +18,176 @@ expression: get_spans()
...
+              "module_path": "apollo_router::plugins::telemetry",
+              "fields": {
+                "names": [
+                  "otel.kind"
+                ]
+              }
+            }
+          },
+          "children": {
+            "apollo_router_core::query_planner::fetch": {
```
Comment (BrynCooke) [inline apollo-router/tests/snapshots/integration_tests__traced_basic_request.snap:131]: Shoudl these also have kind?

**Review by BrynCooke — APPROVED** [review]
Some minor comments.

**Pull request merged**